week1/materials/complexity_analysis.md

```diff
@@ -0,0 +1,150 @@
+of instructions will be `7 + 6n`.
+
+So we see that this formula depends on the machine that we run the program on.
+Isn't there a way to generalize the formula so that it holds true regardless of
+the computational model used for executing the program?
+
+## Asymptotic complexity
+
+Processors, memory access time, language interpretors, virtual machines could
+differ. The total running time of our algorithms depends on all those factors.
+It is hard to predict the total running time of a function before executing it.
+
+However, what we could do easily is tell how a function behaves when its input
+changes. What is the behaviour of `average` when we run it with list of ten numbers?
+`6 + 5 * 10 = 56`. 56 instructions. What about a list of hundred elements?
+`6 + 5 * 100 = 506`. Finally, lets look at a thousand. `6 + 5 * 1000 = 5006`.
+
+What we observe is that the constant `6` is quickly swallowed by the other terms.
+`5n` is the dominating term. The whole function changes in a similar manner to the
+way `5n` changes.
+
+This is the big idea of asymptotic analysis. We are allowed to remove terms that
+are not contributing much to the value of a function as its input grows. Similarly
+to the way we dropped `6` we could also remove the constant in front of `n`. This
+way we could say that as the input to the function changes, `average` is similar
+to `n` (it is linear). Thus, we say `average` is `O(n)`.
+
+[Asymptotic computational complexity](http://en.wikipedia.org/wiki/Asymptotic_computational_complexity)
+
+[Big O notation](http://en.wikipedia.org/wiki/Big_O_notation)
+
+## Amortized complexity
+
+[Amortized complexity](http://en.wikipedia.org/wiki/Amortized_analysis) looks at
+the complexity of a batch of operations (not a single operation). For example,
+Vector's worst case complexity for adding an element is O(n). However, the
+amortized complexity for adding `n` elements to a Vector is O(1).
+
+## Best/average/worst analysis
+
+Lets look at another example - a function that checks if there is an even number
+in a sequence.
+
+```
+has_even(numbers) {
+  n = length(numbers)
+
+  i = 0
+  while (i < n) {
+    if (numbers[i] % 2 == 0) {
+      return true
+    }
+    i = i + 1
+  }
+
+  return false
+}
+```
+
+What is the running time of `has_even`? If first number is even it breaks immediately
+so it is O(1). However, if there aren't any even numbers it will go through all elements
+increasing the running time to O(n).
+
+We say that the best case performance of `has_even` is O(1). This means that given
+the best possible input (even element in the beginning of the sequence) the function
+runs in O(1) time.
+
+We say that the worst case performance of `has_even` is O(n). This means that given
+the worst possible input (even element is not present in the sequence) the function
+runs in O(n) time.
+
+On average the function is O(n) because we have to scan every element until we find
```

Comment:
"we have to scan every element" is an explanation of the worst case. To get the average, you need to analyze the different positions where the first even number can be in - so the calculation is something like 1/n * O(1) + 1/n * O(2) + ... + 1/n * O(n) = n * (n + 1) / (2 * n) = O(n / 2) = O(n).

Also, if the data is uniformly distributed over some range, I'm pretty sure the average complexity is O(1) since there is a 1/2 chance of the first number being even, etc. But I guess you can spare that in class :)